lib/sudo-exec/src/pty.rs

```diff
@@ -28,55 +28,69 @@ impl PtyRelay {
         sudo_pid: ProcessId,
         pty_leader: OwnedFd,
         mut backchannel: ParentBackchannel,
-    ) -> io::Result<Self> {
-        let signal_handlers = SignalHandlers::new()?;
+    ) -> io::Result<(Self, EventHandler<Self>)> {
+        let mut event_handler = EventHandler::<Self>::new()?;
+
+        macro_rules! set_signal_handler {
+            ($($signo:ident,)*) => {
+                $({
+                    event_handler.set_signal_callback::<$signo>(|ec, ev| {
+                        if let Ok(info) = ev.recv_signal_info::<$signo>().unwrap() {
+                            ec.relay_signal(info, ev);
+                        }
+                    })
+                })*
+            };
+        }
+
+        set_signal_handler! {
+            SIGINT, SIGQUIT, SIGTSTP, SIGTERM, SIGHUP, SIGALRM, SIGPIPE, SIGUSR1, SIGUSR2,
+            SIGCHLD, SIGCONT, SIGWINCH,
+        }
+
+        event_handler.set_read_callback(&backchannel, |ec, ev| ec.check_backchannel(ev));
 
         retry_while_interrupted(|| backchannel.send(MonitorEvent::ExecCommand))?;
 
-        Ok(Self {
-            signal_handlers,
-            monitor_pid,
-            sudo_pid,
-            command_pid: None,
-            _pty_leader: pty_leader,
-            backchannel,
-        })
+        Ok((
+            Self {
+                monitor_pid,
+                sudo_pid,
+                command_pid: None,
+                _pty_leader: pty_leader,
+                backchannel,
+                last_event: None,
+            },
+            event_handler,
+        ))
     }
 
-    pub(super) fn run(mut self) -> io::Result<(ExitReason, impl FnOnce())> {
-        loop {
-            // First we check the backchannel for any status updates from the command or the
-            // monitor.
-            if let ControlFlow::Break(event) = self.check_backchannel() {
-                let exit_reason = match event {
-                    ParentEvent::CommandExit(code) => ExitReason::Code(code),
-                    ParentEvent::CommandSignal(signal) => ExitReason::Signal(signal),
-                    ParentEvent::IoError(raw) => return Err(io::Error::from_raw_os_error(raw)),
-                    // We never break the event loop because of this event.
-                    ParentEvent::CommandPid(_) => unreachable!(),
-                };
-
-                return Ok((exit_reason, move || drop(self.signal_handlers)));
-            }
+    pub(super) fn run(mut self, event_handler: &mut EventHandler<Self>) -> io::Result<ExitReason> {
+        event_handler.event_loop(&mut self);
+        // The event loop is only broken if `last_event` is set.
+        let exit_reason = match self.last_event.unwrap() {
```

Comment:
to avoid this `unwrap` maybe the `EventHandler::set_break` could take a value that's returned by `EventHandler::event_loop`. the `brk` field would change its type to `Option<U>` and `set_break` would have signature `fn(&mut self, U)`

that would you would not need the `PtyRelay.last_event` field. when no value needs to be returned from the `event_loop` you can return the `()` type

Reply:
I like that! Maybe I could use `std::ops::ControlFlow`.